core/resharding.go

```diff
@@ -138,33 +144,40 @@ func GetShardingStateFromBlockChain(bc *BlockChain, epoch uint64) *ShardingState
 }
 
 // CalculateNewShardState get sharding state from previous epoch and calculate sharding state for new epoch
-// TODO: currently, we just mock everything
 func CalculateNewShardState(bc *BlockChain, epoch uint64, stakeInfo *map[common.Address]*structs.StakeInfo) types.ShardState {
 	if epoch == GenesisEpoch {
 		return GetInitShardState()
 	}
 	ss := GetShardingStateFromBlockChain(bc, epoch-1)
+	if epoch == FirstEpoch {
+		newNodes := []types.NodeID{}
+		for addr, stakeInfo := range *stakeInfo {
+			newNodes = append(newNodes, types.NodeID{addr.Hex(), hex.EncodeToString(stakeInfo.BlsAddress[:])})
```

Comment:
what's the difference of these two addresses? I thought we can blsAddr as NodeID

Reply:
each node have two addresses. It's better to keep track of both of them.